### PR TITLE
feat: Implement dataset upgrade functionality

### PR DESCRIPTION
**Pull request by pjwerneck**
## Summary

A data owner can now call `client.upgrade()` and have every dataset it owns
published in the protocol layout of the installed client.

## Changes

- Added `DatasetUpgrade` and `UpgradeReport` classes to manage dataset upgrades.
- Implemented `upgrade()` method in `SyftRDSClient` to promote datasets to the current layout.
- Enhanced protocol version handling in dataset creation and sharing processes.
- Updated tests to cover new upgrade functionality, including edge cases for dataset sharing and audience recovery.
- Refactored existing tests to align with the new protocol versioning logic.
- Ensured backward compatibility with older protocol versions during dataset upgrades.

## Testing

New test cases

## Asana task

https://app.asana.com/1/1185126988600652/project/1210542925864934/task/1216142771422684?focus=true

### CODE REVIEW

File: packages/syft-datasets/src/syft_datasets/dataset_manager.py [hunk 1]
```diff
@@ -141,15 +143,62 @@ def migrate(
...
+        ``recover_audience_from_ruleset``, which covers the explicit grants; the
+        "any" flag lives on the transport, so a caller that shares collections
+        must check that too.
         """
         ref = self.storage.find_dataset_ref(self.syftbox_config.email, name)
         migrated = self.storage.migrate_dataset(ref, to_version)
         self._set_new_dataset_permissions(dataset=migrated, users=users)
         return migrated
 
+    def recover_audience_from_ruleset(
```
Comment (rasswanth-s) [inline packages/syft-datasets/src/syft_datasets/dataset_manager.py:157]: great 👍

**Review by rasswanth-s — APPROVED** [review]
Great work 👍
